plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,605 @@
+        if (offset_h.rank() != 4) {
+          offset_h = offset_h_dg;
+        } else {
+          offset_h = op::concatenate({offset_h, offset_h_dg}, 1);
+        }
+        if (offset_w.rank() != 4) {
+          offset_w = offset_w_dg;
+        } else {
+          offset_w = op::concatenate({offset_w, offset_w_dg}, 1);
+        }
+      }
+      // Define height index of input.
+      TensorShape shape_h(precision, {NEW_H, NEW_W});
+      Buffer buffer_h(shape_h);
+      std::vector<float> data_h;
+      for (auto off_h = 0; off_h < OFF_H; ++off_h) {
+        for (auto f_h = 0; f_h < F_H; ++f_h) {
+          auto data = off_h * strides[0] + f_h * dilations[0] - pad_befores[0];
+          for (auto new_w = 0; new_w < NEW_W; ++new_w) {
+            data_h.push_back(static_cast<float>(data));
+          }
+        }
+      }
+      buffer_h.copy_from(data_h.data());
+      auto input_index_h = Constant(buffer_h, "input_index_h");
+      input_index_h = op::reshape(input_index_h, make_tuple<int64_t>({1, 1, NEW_H, NEW_W}));
+      // Define width index of input.
+      TensorShape shape_w(precision, {NEW_W, NEW_H});
+      Buffer buffer_w(shape_w);
+      std::vector<float> data_w;
+      for (auto off_w = 0; off_w < OFF_W; ++off_w) {
+        for (auto f_w = 0; f_w < F_W; ++f_w) {
+          auto data = off_w * strides[1] + f_w * dilations[1] - pad_befores[1];
+          for (auto new_h = 0; new_h < NEW_H; ++new_h) {
+            data_w.push_back(static_cast<float>(data));
+          }
+        }
+      }
+      buffer_w.copy_from(data_w.data());
+      auto input_index_w = Constant(buffer_w, "input_index_w");
+      input_index_w = op::transpose(input_index_w);
+      input_index_w = op::reshape(input_index_w, make_tuple<int64_t>({1, 1, NEW_H, NEW_W}));
+      // Get deformabled index.
+      edsl::Tensor new_index_h = offset_h + input_index_h;
+      edsl::Tensor new_index_w = offset_w + input_index_w;
+      // Get deformable input tensor.
+      edsl::Tensor deform_input = edsl::gather(I, new_index_h).axis(-2);
+      deform_input = edsl::gather(deform_input, new_index_w).axis(-1);
+      deform_input = extract_tensor(deform_input, rank - 2);
+      // Compute DeformableConvolution.
+      autopad_mode = plaidml::op::AutoPadMode::VALID;
+      auto result = op::convolution(deform_input, F)
+                        .strides(std::vector<size_t>({(size_t)F_H, (size_t)F_W}))
+                        .autopad_mode(autopad_mode)
+                        .input_layout(plaidml::op::TensorLayout::NCX)
+                        .filter_layout(plaidml::op::TensorLayout::KCX)
+                        .autogroup_mode(plaidml::op::AutoGroupMode::EXPLICIT)
+                        .group_layout(plaidml::op::GroupLayout::IN_C)
+                        .group_layout(plaidml::op::GroupLayout::IN_K)
+                        .groups(static_cast<int>(G));
+      return edsl::make_tuple(result);
+    }
+    // Compute 3D DeformableConvolution.
```

Comment:
I made some comments for the 2D version, which may apply to 1D and 3D as well, please check. 
It would be great if we can use a single function with parameter `rank`  for 1D, 2D and 3D.
If that is too difficult, we can refactor the code with 3 different functions.